### PR TITLE
Discord Login

### PR DESCRIPTION
**Pull request by timolegros**
## Description
Enables Discord OAuth 2.0 login. It works exactly the same way as Github OAuth 2.0 login.
Introduced 3 new routes: 'discordAccount', '/auth/discord' and '/auth/discord/callback' all of which work the same way as their GitHub counterparts (routes trigger the same functions).

## Motivation and Context
Paves the way for more Discord integrations (sharing, embedding, etc).

## How has this been tested?
Tested by logging in and out locally and on staging-2 using Discord OAuth2.

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

## Does this PR affect any server routes?
- [ ] yes, and they are tested: [enter the % coverage here]
- [x] yes, and they are not tested
- [ ] yes, but I did not run tests
- [ ] no

### CODE REVIEW

**Comment by zakhap** [comment]
@timolegros @alexyoung23j Is this PR ready for review?

File: server/passport.ts [hunk 1]
```diff
@@ -251,6 +253,11 @@ function setupPassport(models: DB) {
     callbackURL: GITHUB_OAUTH_CALLBACK,
     passReqToCallback: true,
   }, async (req: Request, accessToken, refreshToken, profile, cb) => {
+    const str = '&state='
```
Comment (jnaviask) [inline server/passport.ts:256]: What's this session validation do for Github?

File: server/passport.ts [hunk 2]
```diff
@@ -348,6 +353,116 @@ function setupPassport(models: DB) {
       return cb(null, newUser);
     }
   }));
+
+  if (DISCORD_CLIENT_ID && DISCORD_CLIENT_SECRET && DISCORD_OAUTH_CALLBACK) passport.use(new DiscordStrategy({
```
Comment (jnaviask) [inline server/passport.ts:357]: Could we factor this + github out, to make it easy to add new SSO logins in the future using the same logic? Or are the two too different?

File: README.md [hunk 3]
```diff
@@ -95,6 +95,9 @@ Environment variables used for external services include:
 - MAGIC_API_KEY
 - MAGIC_SUPPORTED_BASES
 - MAGIC_DEFAULT_CHAIN
+- DISCORD_CLIENT_ID: for Discord OAuth login
+- DISCORD_CLIENT_SECRET: for Discord OAuth login
+- DISCORD_OAUTH_SCOPES: scopes (usually just 'identify')
```
Comment (jnaviask) [inline README.md:100]: This variable unused?

**Review by jnaviask — COMMENTED** [review]
A couple small comments -- otherwise looks good. Did not QA.

File: server/passport.ts (hunk 1 above)
Comment (timolegros) [reply server/passport.ts:256]: It's simply a security feature. You pass it in when you start the authentication process and GitHub will add the state param to the callback URL.

File: README.md (hunk 3 above)
Comment (timolegros) [reply README.md:100]: Not sure what you mean they are all being used.

File: server/passport.ts (hunk 2 above)
Comment (timolegros) [reply server/passport.ts:365]: fixed

File: README.md (hunk 3 above)
Comment (jnaviask) [reply README.md:100]: Can you explain what scopes does?

Comment (timolegros) [reply README.md:100]: It's just the permissions that a user agrees to give us (in terms of access to their account). So for example, "identify" allows us to get their username/profile info.

**Pull request merged**